crates/bevy_ecs/src/query/iter.rs

```diff
@@ -849,12 +913,27 @@ impl<'w, 's, D: QueryData, F: QueryFilter> Iterator for QueryIter<'w, 's, D, F>
                 let archetype =
                     // SAFETY: Matched archetype IDs are guaranteed to still exist.
                     unsafe { self.archetypes.get(id.archetype_id).debug_checked_unwrap() };
-                accum =
+                // SAFETY: Matched table IDs are guaranteed to still exist.
+                let table = unsafe { self.tables.get(archetype.table_id()).debug_checked_unwrap() };
+
+                // When an archetype and its table have equal entity counts, dense iteration can be safely used.
+                // this leverages cache locality to optimize performance.
+                if table.entity_count() == archetype.len() {
```

Comment:
```suggestion
                if D::is_dense && table.entity_count() == archetype.len() {
```

It might only be worth checking for this if the filter is the only sparse part of the query. Otherwise we'd be inserting extra branches into queries that will never be able to use dense iteration